ui/src/addCriteria.tsx

```diff
@@ -0,0 +1,202 @@
+  const source = useSource();
+  const navigate = useNavigate();
+  const dispatch = useAppDispatch();
+  const { cohort, group } = useCohortAndGroup();
+
+  const [query, setQuery] = useState<string>("");
+  const [showResults, setShowResults] = useState<boolean>(false);
+  const [data, setData] = useState<TreeGridData>({});
+
+  const criteriaConfigs = underlay.uiConfiguration.criteriaConfigs;
+  const searchConfig = underlay.uiConfiguration.criteriaSearchConfig;
+
+  const categories = useMemo(() => {
+    const categories: CriteriaConfig[][] = [];
+
+    for (const config of criteriaConfigs) {
+      let category: CriteriaConfig[] | undefined;
+      for (const c of categories) {
+        if (c[0].category === config.category) {
+          category = c;
+          break;
+        }
+      }
+
+      if (category) {
+        category.push(config);
+      } else {
+        categories.push([config]);
+      }
+    }
+
+    return categories;
+  }, [underlay]);
+
+  const columns = useMemo(
+    () => [
+      {
+        key: "t_criteria_type",
+        width: searchConfig.criteriaTypeWidth,
+        title: "Criteria type",
+      },
+      ...searchConfig.columns,
+    ],
+    [underlay]
+  );
+
+  const onClick = useCallback(
+    (config: CriteriaConfig, dataEntry?: DataEntry) => {
+      const criteria = createCriteria(source, config, dataEntry);
+      dispatch(
+        insertCriteria({
+          cohortId: cohort.id,
+          groupId: group.id,
+          criteria,
+        })
+      );
+      navigate(
+        !!getCriteriaPlugin(criteria).renderEdit && !dataEntry
+          ? "../" + criteriaURL(criteria.id)
+          : "../../" + cohortURL(cohort.id, group.id)
+      );
+    },
+    [source, cohort.id, group.id]
+  );
+
+  const criteriaConfigMap = useMemo(
+    () => new Map(criteriaConfigs.map((cc) => [cc.id, cc])),
+    [criteriaConfigs]
+  );
+
+  const search = useCallback(async () => {
+    // This prevents briefly showing the empty results from the empty query
+    // inbetween when a new query has been set but before searchState is
+    // retriggered.
+    setShowResults(!!query);
+
+    const children: DataKey[] = [];
+    const data: TreeGridData = {
+      root: { data: {}, children },
+    };
+
+    if (query) {
+      const res = await searchCriteria(source, criteriaConfigs, query);
+
+      res.data.forEach((entry) => {
+        const key = `${entry.source}~${entry.data.key}`;
+        children.push(key);
+
+        const item: CriteriaItem = {
+          data: {
+            ...entry.data,
+            t_criteria_type: (
+              <Stack direction="row" justifyContent="center">
+                <Chip
+                  label={criteriaConfigMap.get(entry.source)?.title}
+                  size="small"
+                />
+              </Stack>
+            ),
+          },
+          entry: entry,
+        };
+        data[key] = item;
+      });
+    }
+
+    setData(data);
+  }, [source, query]);
```

Comment:
should this not also depend on `criteriaConfigs` (and possibly `criteriaConfigMap`, though I think that's redundant)?

Reply:
Done. It doesn't actually matter because the underlays are never changed without reloading but it's good to be correct to prevent future problems.